### ISSUE TITLE
don't have dirty low card ints - may result in too much continuous?

### ISSUE THREAD

**Issue by amueller** [dabl/dabl#163]
if something is a mix of integers and strings it'll become a dirty float if it has more than 5 integer values even though it could be categorical.
Doing this right would complicate the logic *a lot* but seems like a big issue.
```python
import numpy as np
import pandas as pd
from dabl import detect_types

df = pd.DataFrame({'a': np.random.randint(0, 10, size=1000)})
print(detect_types(df))
# low_card_int

df.a[0] = 'a'
print(detect_types(df))
# dirty_float
```

**Comment by kumanday** [comment]
That `5` rang a bell, so I checked.. and this is apparently indeed a result of the hardcoded `5` in `_find_string_floats`
<img width="574" alt="Screen Shot 2020-01-31 at 2 46 35 PM" src="https://user-images.githubusercontent.com/1234749/73569581-b85a3b80-4438-11ea-8d42-c49232de5f82.png">
When I change it to `8`, the behavior changes accordingly.

What would be the desired behavior?

**Comment by amueller** [comment]
I think the ideal behavior would be to figure out that it's an integer and not a float. Integers go through a different route for determining whether something is a categorical or a continuous variable (via a bunch of hacks).
I'm not sure if introducing a new ``dirty_low_card_int`` type is what makes most sense. Basically the preprocessing both keeps the original and adds one-hot-encoding for low_card_int. In plotting, there's a heuristic that looks at the smoothness of the histogram to decide whether the variable should be treated as categorical or ordinal.
If anything, if a variable is a mix of ints and strings, I would expect it to be more likely to be categorical than if it was only ints, not the other way around...

Do you have an example where this happens?

**Comment by kumanday** [comment]
I agree with the premise.  I've worked a lot with CCN identifiers (CMS Certification Number, from Centers for Medicare and Medicaid Services) that are categorical but unfortunately > 99% integer-like, with a few including leading zeros or a single letter.  I've suffered with ETL pipelines and spreadsheets that mangle the type, strip out the leading 0, etc.

On the other hand, I've other government-provided numerical datasets that have a few ill-conceived ideas for "missing value" identifiers in the form of strings, such as `''`, `'(blank)'`, `'(not available)'`, or `'(missing)'`.  Perhaps we could first try a look-up of possible missing-value strings before designating the data type as categorical.

**Comment by kumanday** [comment]
Speaking of leading zeros, I have also suffered with datasets treating zip codes as integers -- that doesn't work well for Cambridge, MA. ;-)   This might be a common enough use case that it warrants a special consideration of possibly being a zip code for sets of 5-digit "integers" that include some with a leading zero.

**Comment by amueller** [comment]
That's the reason why there is a "low_card_integer" type: it's really hard to detect zip-codes etc.
The goal of the "dirty float" was to ensure that if there's some beautiful encoding of missing values that we'll capture that. The hard-coded 5 is based on me expecting that there are not more than 5 different ways to specify "missing value" in a weird string (for each column).

The "dirty_low_card_int" basically would again try to remove the very common non-integer values (assuming they are some encoding of "missing" / "na" etc.) and then handle the rest as "low_card_int" which can later be treated by a heuristic.

Right now, if you have a "clean" zip code, I assume the plotting will figure out it's categorical because the histogram will not be smooth. The point of this issue was that basically I want this heuristic to kick in on the things that are not strings.

All of this clearly needs more documentation....

**Comment by amueller** [comment]
Related / duplicate of #69

**Comment by amueller** [comment]
Fixed by #270.
